extensions-core/kinesis-indexing-service/src/main/java/org/apache/druid/indexing/kinesis/supervisor/KinesisSupervisor.java

```diff
@@ -314,4 +356,94 @@ protected boolean useExclusiveStartSequenceNumberForNonFirstSequence()
   {
     return true;
   }
+
+  @Override
+  protected Map<String, OrderedSequenceNumber<String>> filterDeadShardsFromStartingOffsets(
+      Map<String, OrderedSequenceNumber<String>> startingOffsets
+  )
+  {
+    Map<String, OrderedSequenceNumber<String>> filteredOffsets = new HashMap<>();
+    for (Map.Entry<String, OrderedSequenceNumber<String>> entry : startingOffsets.entrySet()) {
+      if (!entry.getValue().get().equals(KinesisSequenceNumber.END_OF_SHARD_MARKER)) {
+        filteredOffsets.put(entry.getKey(), entry.getValue());
+      } else {
+        log.info("Excluding shard[%s] because it has reached EOS.", entry.getKey());
+      }
+    }
+    return filteredOffsets;
+  }
+
+  @Override
+  protected void cleanupDeadShards(Set<String> expiredShards)
+  {
+    log.info("Cleaning up dead shards: " + expiredShards);
+
+    final KinesisDataSourceMetadata dataSourceMetadata =
+        (KinesisDataSourceMetadata) getIndexerMetadataStorageCoordinator().getDataSourceMetadata(dataSource);
```

Comment:
The introduction of `getIndexerMetadataStorageCoordinator()` seems like an abstraction-break, in that the parent class is meant to handle interactions with the metadata store and the subclass is meant to handle interactions and customizations relevant to the upstream system.

It is used for two things:

- Getting the currently-committed datasource metadata.
- Later on, possibly resetting it.

These actions when done outside of the parent class, raise some alarms about possible racey behavior.

Maybe the currently-committed datasource metadata could be passed into `cleanupDeadShards`? And it could return new metadata, which the parent class could commit after doing some sanity checks?

Reply:
Adjusted this to follow the suggested interaction between the subclass and parent class.

> after doing some sanity checks

Did you have some specific sanity checks in mind?

Reply:
> Did you have some specific sanity checks in mind?

I was thinking stuff like: the new partition set is a subset of the old one, and offsets haven't changed (the method is named `cleanupDeadShards` so it shouldn't be changing offsets or adding partitions).

Reply:
Thanks, added a sanity check there